### DISCUSSION TITLE
Feature Proposal: Observability/Metrics APIs.

### DISCUSSION THREAD

**Discussion by jyu283** [jyu283/gradescope.py#41] [Ideas]
Right now, @DuckDuckWhaleUCSD has built an observer wrapper around the autograder to capture grading output to a server. If my understanding is correct, this system relies on the logs emitted by the tritongrader library to function properly. There's also a problem with gradescope JSON files being huge. (I'm not too familiar with this issue, @DuckDuckWhaleUCSD can expand on this further in the comments.)

To make the process more streamlined, I can try to move some of the APIs into this library. The rough idea is that in addition to formatter classes such as `GradescopeResultsFormatter`, we can also instantiate a class that transmits formatted grading results to a specified endpoint. Right now this endpoint is being managed by @DuckDuckWhaleUCSD on a UCSD server.

**Comment by DuckDuckWhaleUCSD** [comment]
Currently the logs are only saved in case the autograder returns with a non-zero exit code and will be sent to Slack as the error message. (We should work out how to forward only the error part.)

The Gradescope metadata JSON is stupid big (>10 MB I think now? I need to check) as it includes all outputs from all previous submissions so I'm thinking the truly future-proof solution is to enable zstd compression on the body. That means probably adding a Linux python-zstd-something package on the setup script and adding a dep in Observer.

The new endpoint for the stats need to see the actual score before any hiding is done. Either we go with Gradescope JSON format (need to look into this) but it could be complicated. Or we can make a simpler format with probably JSON that is just `[{name: "Test 1", score: 10, max: 20}]`. I don't really see a point of transmitting per test case, that creates more load on the server (even though it's stupidly powerful so that hardly matters), more points of failure, doesn't work well with our seccomp hardening (no socket syscalls during grading), so probably a final JSON file dump that get picked up by the end wrapper Python script would work.
